### PR TITLE
HDDS-6845. Add CI to check pull request title

### PR DESCRIPTION
**Pull request by kaijchen**
## What changes were proposed in this pull request?

Enforce the PR title to match the pattern `^HDDS-[1-9][0-9]{3,4}\. .+$`. 

```
HDDS-1234. Pull Request Title    // ok
HDDS-1234 Pull Request Title     // bad (missing dot)
HDDS-1234.Pull Request Title     // bad (missing space)
HDDS-0123. Pull Request Title    // bad (leading zero)
HDDS-12345. Pull Request Title   // ok (ready for 5 digit Jira number)
HDDS-123456. Pull Request Title  // bad (to be enabled when needed)
```

EDIT: it is also able to detect 2 consecutive spaces and trailing space, such as:

```
// please ignore the quote
"HDDS-1234. Pull Request Title " // bad (trailing space)
"HDDS-1234.  Pull Request Title" // bad (double space)
```

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6845

## How was this patch tested?

Case coverage: `dev-support/ci/pr_title_check.bats`
In action: https://github.com/apache/ozone/actions/workflows/pull-request.yml

### CODE REVIEW

**Comment by kaijchen** [comment]
Closed due to `slashgear/action-check-pr-title@v3.0.0 is not allowed to be used in apache/ozone.`

File: .github/workflows/pull-request.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+jobs:
+  title:
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Check pull request title
+        env:
+          TITLE: ${{ github.event.pull_request.title }}
+        run:
+          echo ${TITLE} | grep -E '^HDDS-[1-9][0-9]{3,4}\. .+$'
```
Comment (sadanand48) [inline .github/workflows/pull-request.yml:34]: `"HDDS-1234.  Pull Request Title "`. multiple spaces after dot is a bad case

**Review by sadanand48 — COMMENTED** [review]
Thanks for working on this @kaijchen . In some cases, the pr title has the feature label like [FSO] or [EC]. I guess this won't get handled here

**Comment by kaijchen** [comment]
> Thanks for working on this @kaijchen . In some cases, the pr title has the feature label like [FSO] or [EC]. I guess this won't get handled here

I think those cases should be fine.
For example: `HDDS-6708. [FSO] Make normalize keyPath on key rename request (#3386)`.

File: .github/workflows/pull-request.yml (hunk 1 above)
Comment (kaijchen) [reply .github/workflows/pull-request.yml:34]: Thanks for the advice.

**Comment by sadanand48** [comment]
> I think those cases should be fine.

Sure thanks for clarifying.

File: .github/workflows/pull-request.yml (hunk 1 above)
Comment (kaijchen) [reply .github/workflows/pull-request.yml]: Covered now: https://github.com/apache/ozone/runs/6792860766?check_suite_focus=true

**Comment by kaijchen** [comment]
Trailing space is covered in https://github.com/apache/ozone/runs/6792930448?check_suite_focus=true

**Comment by kaijchen** [comment]
Thanks for the review @sadanand48, I have updated this PR to cover cases with illegal spaces.
Please take another look.

PS: Do you think we should check the trailing `.`?

File: .github/workflows/pull-request.yml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+jobs:
+  title:
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Check pull request title
+        env:
+          TITLE: ${{ github.event.pull_request.title }}
+        run:
+          echo "${TITLE}" | grep -E '^HDDS-[1-9][0-9]{3,4}\. .*[^ ]$'
```
Comment (sadanand48) [inline .github/workflows/pull-request.yml:34]: We could also merge both the checks into one using pipe ,` echo "${TITLE}" | grep -E '^HDDS-[1-9][0-9]{3,4}\. .*[^ ]$' | grep -v '   '`

Comment (kaijchen) [reply .github/workflows/pull-request.yml:34]: good idea, thanks

**Comment by sadanand48** [comment]
Thanks for updating the PR.

> : Do you think we should check the trailing `.`?

I'm not really sure, On checking the current PR Queue, there are a few with '.' and few without it. I think we can leave it

File: .github/workflows/pull-request.yml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+jobs:
+  title:
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Check pull request title
+        env:
+          TITLE: ${{ github.event.pull_request.title }}
+        run:
+          echo "${TITLE}" | grep -E '^HDDS-[1-9][0-9]{3,4}\. .*[^ ]$' | grep -v '  '
```
Comment (sodonnel) [inline .github/workflows/pull-request.yml:34]: If you want to ensure there is not a double space after the `HDDS-1234.` should we add in something explicit, like:

```
^HDDS-[1-9][0-9]{3,4}\.[[:space:]][^[:space:]]+.*[^[:space:]]$
```

Ie, it is a space; zero or 1 "not a space"; zero or more of any character.

**Review by adoroszlai — COMMENTED** [review]
I would suggest extracting the check into a script in `dev-support/ci`, and adding `bats` test for it.

File: .github/workflows/pull-request.yml (hunk 3 above)
Comment (kaijchen) [reply .github/workflows/pull-request.yml:34]: The current expression is checking no double space at any place, and I think it's better (and clearer).

**Comment by kaijchen** [comment]
> I would suggest extracting the check into a script in `dev-support/ci`, and adding `bats` test for it.

It could be in a script but notice the event trigger this CI is different than the others.
We may need to modify `build-info` regarding to this.

**Comment by adoroszlai** [comment]
> > I would suggest extracting the check into a script in `dev-support/ci`, and adding `bats` test for it.
> 
> It could be in a script but notice the event trigger this CI is different than the others. We may need to modify `build-info` regarding to this.

Keeping it in separate workflow is OK.  The test is useful because it captures valid and invalid examples, and is easier to run than manually editing PR titles.

**Comment by kaijchen** [comment]
> Keeping it in separate workflow is OK. The test is useful because it captures valid and invalid examples, and is easier to run than manually editing PR titles.

I got your point. Thanks for the advice.

File: .github/workflows/pull-request.yml (hunk 3 above)
Comment (sodonnel) [reply .github/workflows/pull-request.yml:34]: Ah ok. I didn't realize the check was for double spaces everywhere. Makes sense to keep it as it is then.

**Comment by kaijchen** [comment]
confirmed bats checked in
https://github.com/apache/ozone/runs/6818147594?check_suite_focus=true#step:5:39

**Comment by adoroszlai** [comment]
Thanks @kaijchen for updating the patch.

The only concern I have is about the requirement for 4-5 digits: there are 93 open Jira issues preceding HDDS-1000.  We could not create PRs for these due to to few digits.  How about a less strict check for any number of digits (or maybe 1-5)?

**Comment by kaijchen** [comment]
> Thanks @kaijchen for updating the patch.
> 
> The only concern I have is about the requirement for 4-5 digits: there are 93 open Jira issues preceding [HDDS-1000](https://issues.apache.org/jira/browse/HDDS-1000). We could not create PRs for these due to to few digits. How about a less strict check for any number of digits (or maybe 1-5)?

Sure, I didn't realize some old Jiras are still open.

**Review by adoroszlai — APPROVED** [review]

**Comment by adoroszlai** [comment]
Thanks @kaijchen for the contribution, @sadanand48 and @sodonnel for the reviews.

**Comment by kaijchen** [comment]
Thanks @sadanand48 @sodonnel @adoroszlai for the reviews.

**Pull request merged**